### ISSUE TITLE
twisted.trial.test.test_assertions.TestSynchronousAssertions.test_assertRaisesContextUnexpected fails on Python 3.3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7534]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#7534|
|Type|release blocker: regression|
|Created|2014-07-02 13:07:45Z|
|Branch|https://github.com/twisted/twisted/tree/assertraises-py3-7534|

Maybe only sometimes?  It's unclear.

```
======================================================================
FAIL: test_assertRaisesContextUnexpected (twisted.trial.test.test_assertions.TestSynchronousAssertions)
test_assertRaisesContextUnexpected
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/buildslave/run/python-3.3-tests/Twisted/twisted/trial/test/test_assertions.py", line 415, in test_assertRaisesContextUnexpected
    raise TypeError('marker')
TypeError: marker

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/buildslave/run/python-3.3-tests/Twisted/twisted/trial/test/test_assertions.py", line 415, in test_assertRaisesContextUnexpected
    raise TypeError('marker')
  File "/home/buildslave/run/python-3.3-tests/Twisted/twisted/trial/_synctest.py", line 359, in __exit__
    self._expectedName, reason.getTraceback()),
  File "/home/buildslave/run/python-3.3-tests/Twisted/twisted/trial/_synctest.py", line 381, in fail
    raise self.failureException(msg)
twisted.trial.unittest.FailTest: builtins.TypeError raised instead of ValueError:
 Traceback (most recent call last):
  File "/home/buildslave/run/python-3.3-tests/Twisted/twisted/trial/_synctest.py", line 1339, in _runFixturesAndTest
    if self._run(suppress, todo, method, result):
  File "/home/buildslave/run/python-3.3-tests/Twisted/twisted/trial/_synctest.py", line 1308, in _run
    runWithWarningsSuppressed(suppress, method)
  File "/home/buildslave/run/python-3.3-tests/Twisted/twisted/python/util.py", line 1021, in runWithWarningsSuppressed
    return f(*args, **kwargs)
  File "/home/buildslave/run/python-3.3-tests/Twisted/twisted/trial/test/test_assertions.py", line 415, in test_assertRaisesContextUnexpected
    raise TypeError('marker')
--- <exception caught here> ---
  File "/home/buildslave/run/python-3.3-tests/Twisted/twisted/trial/test/test_assertions.py", line 415, in test_assertRaisesContextUnexpected
    raise TypeError('marker')
builtins.TypeError: marker


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/buildslave/run/python-3.3-tests/Twisted/twisted/trial/_synctest.py", line 1308, in _run
    runWithWarningsSuppressed(suppress, method)
  File "/home/buildslave/run/python-3.3-tests/Twisted/twisted/python/util.py", line 1021, in runWithWarningsSuppressed
    return f(*args, **kwargs)
  File "/home/buildslave/run/python-3.3-tests/Twisted/twisted/trial/test/test_assertions.py", line 423, in test_assertRaisesContextUnexpected
    "{0}".format(message))
  File "/home/buildslave/run/python-3.3-tests/Twisted/twisted/trial/_synctest.py", line 403, in assertTrue
    raise self.failureException(msg)
twisted.trial.unittest.FailTest: Exception message did not begin with expected information: builtins.TypeError raised instead of ValueError:
 Traceback (most recent call last):
  File "/home/buildslave/run/python-3.3-tests/Twisted/twisted/trial/_synctest.py", line 1339, in _runFixturesAndTest
    if self._run(suppress, todo, method, result):
  File "/home/buildslave/run/python-3.3-tests/Twisted/twisted/trial/_synctest.py", line 1308, in _run
    runWithWarningsSuppressed(suppress, method)
  File "/home/buildslave/run/python-3.3-tests/Twisted/twisted/python/util.py", line 1021, in runWithWarningsSuppressed
    return f(*args, **kwargs)
  File "/home/buildslave/run/python-3.3-tests/Twisted/twisted/trial/test/test_assertions.py", line 415, in test_assertRaisesContextUnexpected
    raise TypeError('marker')
--- <exception caught here> ---
  File "/home/buildslave/run/python-3.3-tests/Twisted/twisted/trial/test/test_assertions.py", line 415, in test_assertRaisesContextUnexpected
    raise TypeError('marker')
builtins.TypeError: marker


----------------------------------------------------------------------
```

<details><summary>Searchable metadata</summary>

```
trac-id__7534 7534
type__release_blocker__regression release blocker: regression
reporter__exarkun exarkun
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_assertraises_py3_7534 branches/assertraises-py3-7534
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1404306465316366 1404306465316366
changetime__1410710187010151 1410710187010151
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [43062]) Branching to 'assertraises-py3-7534'

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl set owner to @exarkun|
|-|-|

Hi exarkun,

Thanks for fixing this up. Looks good to me, and the test passes on py3.3 now, with the buildbots being one square greener, and the others being unrelated errors. One hasn't finished yet, as of writing (https://buildbot.twistedmatrix.com/builders/fedora17-x86_64-py2.7/builds/2063), but as all the others are green, and it's at 43 minutes right now, I think we're safe.

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [43065]) Merge assertraises-py3-7534

Author: exarkun (sponsored by Hybrid Logic, Ltd.)
Reviewer: hawkowl
Fixes: [#7534](https://github.com/twisted/twisted/issues/7534)

Make the test for the error case of `TestCase.assertRaises` account for the
differing names of `TypeError` on Python 2 vs Python 3.  This fixes the test
on Python 3.
